### DISCUSSION TITLE
Adding a scalar from a library

### DISCUSSION THREAD

**Discussion by mmahalwy** [hayes/pothos#106] [Q&A] [answered]
When using a library such as `graphql-scalars` for the JSON scalar, it's unclear how to add it to Giraphql. I tried this
```ts
const JsonType = builder.scalarType('GraphQLJSON', GraphQLJSON);
```

which resulted to error below:
![image](https://user-images.githubusercontent.com/3749095/118422964-ed63bf00-b678-11eb-83b0-b35f60c05319.png)

Seems like the type needs to change to: 
```ts
description?: string | null;
extensions?: Readonly<Record<string, unknown>> | null;
```

Narrowed it down to this:
https://github.com/hayes/giraphql/blob/1da1283475089311aed09a1c5079007e9abb7abf/packages/core/src/types/global/type-options.ts#L28

**Comment by hayes** [comment, marked as answer]
There are 2 methods:

scalarType let's you define your own scalars by implementing methods directly.

addScalarType let's you add existing scalar definitions.
https://giraphql.com/guide/scalars#adding-graphql-scalars

**Comment by mmahalwy** [comment]
Awesome, thank you @hayes!

**Comment by baconcheese113** [comment]
Trying this with `DateTime` from [graphql-iso-date](https://github.com/excitement-engineer/graphql-iso-date) and it's [TS definitions from DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/graphql-iso-date/index.d.ts) with this line:

```ts
import { GraphQLDateTime } from 'graphql-iso-date'

export const builder = new SchemaBuilder<{
    Context: IContext;
    PrismaTypes: PrismaTypes
    Scalars: {
        DateTime: {
            Input: Date
            Output: Date
        }
    }
}>({
    plugins: [PrismaPlugin],
    prisma: {
        client: prisma,
    },
})
builder.addScalarType("DateTime", GraphQLDateTime, {})
```
and I get the following error
```
Argument of type 'GraphQLScalarType' is not assignable to parameter of type 'GraphQLScalarType<unknown, unknown>'.
Type 'GraphQLScalarType' is missing the following properties from type 'GraphQLScalarType<unknown, unknown>': specifiedByURL, [Symbol.toStringTag]
```
This must be a common scalar type to add, any ideas what I'm doing wrong?

**Reply by hayes** [reply]
I think the type definitions from `graphql-iso-date` might just be out of date, I think this is a case of having multiple versions of the graphql library installed.  Personally I would use https://www.graphql-scalars.dev/docs/scalars/date-time instead.

**Reply by baconcheese113** [reply]
Ah, that worked! I just switched to
```ts
import { GraphQLDateTime } from `graphql-scalars`
```

So after defining that scalar is this the best way to use it with Prisma?
```ts
builder.prismaObject('Event', {
    fields: (t) => ({
        id: t.exposeID('id'),
        sensor: t.relation('sensor'),
        propagatedAt: t.field({
            type: 'DateTime',
            nullable: true,
            resolve: (event) => event.propagatedAt,
        }),
        createdAt: t.field({
            type: 'DateTime',
            resolve: (event) => event.createdAt,
        }),
    })
})
```
Also is there a way to import the descriptions from the prisma schema?

**Reply by hayes** [reply]
> So after defining that scalar is this the best way to use it with Prisma?

Thats probably the most consistent option.  Another option would be to use `propagatedAt: t.expose('propagatedAt', { type: 'DateTime', nullable: true })`.  Either of these should work.

> Also is there a way to import the descriptions from the prisma schema?

Just pushed an option for this,  after upgrading you can enable this with a new `prisma: { exposeDescriptions: true }` option when creating the builder.

Related to the above, t.expose and t.relation, t.relatedConnection, t.prismaObject and t.prismaNode are the methods that can automatically add descriptions from the prisma schema.  so using `propagatedAt: t.expose('propagatedAt', { type: 'DateTime', nullable: true })` would be best if you want descriptions for your date fields
